### PR TITLE
Enable Tailscale's builtin inbound HTTPS proxy

### LINKED ISSUE #62: HTTPS Support
HTTPS Support

https://tailscale.com/kb/1153/enabling-https/

Is it possible to add support for requesting an SSL certificate with Tailscale?

### PR DESCRIPTION
**Pull request by lmagyar**
Replaces PR #89 #134

# Proposed Changes

This is a better solution to the #62 issue than #89 #134 because it doesn't need any manual configuration and also provides a built-in proxy for the https traffic. We don't need to refresh the certificates manually, this is all done by tailscale automatically.

## Related Issues

Fixes #62

### CODE REVIEW

**Comment by lmagyar** [comment]
not stale

**Comment by lmagyar** [comment]
I force pushed a rebase onto the lates PR #133, because with tailscale 1.38.1 some script changes are required. So this PR works only after that is merged.

File: LICENSE.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # MIT License
 
-Copyright (c) 2021-2022 Franck Nijhof
+Copyright (c) 2021-2023 Franck Nijhof
```
Comment (frenck) [inline LICENSE.md:3]: PR is full of unrelated changed.

**Review by frenck — CHANGES_REQUESTED** [review]
The PR references a lot, but it doesn't describe what it does.

**Comment by lmagyar** [comment]
I stashed and rebased it.

This is an optional functionality, if the user doesn't configure anything, it works as previously, but emits some info log messages, that it can be configured and used.

If configured, this solves the unsecure connection problem (I know, it's secure, only the browsers have problem with it): downloads and regularly refreshes a tailscale certificate for our tailnet and fires up it's own built-in proxy.

**Comment by lmagyar** [comment]
I tried to keep this add-on as zero-config as it was, all the configuration is done on the tailscale admin web UI.

File: tailscale/DOCS.md [hunk 2]
```diff
@@ -17,8 +17,55 @@ the following URL:
 
 <https://login.tailscale.com/start>
 
-You can also create an account during the add-on installation processes,
-however, it is nice to know where you need to go later on.
+<details>
+<summary>
```
Comment (frenck) [inline tailscale/DOCS.md:21]: Let's not do those weird collapsable, this is not a 10 pages doc.
The fact you have to write `(click to open)` is already a good indication it is bad ux.

File: tailscale/DOCS.md [hunk 3]
```diff
@@ -17,8 +17,55 @@ the following URL:
...
+<b>(click to open)</b></i>
+</summary>
+
+### Home Assistant configuration for Tailscale Proxy feature
+
+You must configure Home Assistant to **not** use SSL certificates, to be
+accessible through plain http connection. The Tailscale https Proxy will access
+Home Assistant through `localhost` and will not accept a real certificate,
+connection will be closed with `proxy error: x509: cannot validate certificate
+for 127.0.0.1 because it doesn't contain any IP SANs`
```
Comment (frenck) [inline tailscale/DOCS.md:33]: This is the default, I think just mention that Home Assistant itself should not use certificates and link to the official documentation is enough.

File: tailscale/DOCS.md [hunk 4]
```diff
@@ -17,8 +17,55 @@ the following URL:
...
+
+If you still want to use another https connection to access Home Assistant
+through another network, please use the **NGINX Home Assistant SSL proxy**
+add-on.
```
Comment (frenck) [inline tailscale/DOCS.md:37]: Just stating to use a reverse proxy should be enough. I generally like to withhold recommendations of other add-ons.

File: tailscale/DOCS.md [hunk 5]
```diff
@@ -17,8 +17,55 @@ the following URL:
...
+
+So these lines have to be removed from your `/config/configuration.yaml`:
+
+```yaml
+http:
+#  ssl_certificate: /ssl/fullchain.pem
+#  ssl_key: /ssl/privkey.pem
+```
```
Comment (frenck) [inline tailscale/DOCS.md:45]: These are not there by default, the example is even unneeded, as `http` is a default loaded component.

This section can be removed.

File: tailscale/DOCS.md [hunk 6]
```diff
@@ -17,8 +17,55 @@ the following URL:
...
+
+Since Home Assistant by default blocks requests from proxies/reverse proxies,
+you need to tell your instance to allow requests from the Tailscale add-on.
+In order to do so, add the following lines to your `/config/configuration.yaml`
+without changing anything:
```
Comment (frenck) [inline tailscale/DOCS.md:50]:
```suggestion
Since Home Assistant, by default, blocks requests from proxies/reverse proxies,
you need to tell your instance to allow requests from the Tailscale add-on.
In order to do so, add the following lines to your `/config/configuration.yaml`
without changing anything:
```

File: tailscale/DOCS.md [hunk 7]
```diff
@@ -17,8 +17,55 @@ the following URL:
...
+
+```yaml
+http:
+  use_x_forwarded_for: true
+  trusted_proxies:
+    - 127.0.0.1
+```
+
+**Note**: _There is no need to adapt anything in these lines since the addon
```
Comment (frenck) [inline tailscale/DOCS.md:59]: This is already mentioned in the section above. Let's remove it.

File: tailscale/DOCS.md [hunk 8]
```diff
@@ -17,8 +17,55 @@ the following URL:
...
-You can also create an account during the add-on installation processes,
-however, it is nice to know where you need to go later on.
+<details>
+<summary>
+<i>Additional prerequisites if you want to access your Home Assistant device
+with a Tailscale provided certificate, ie. through the Tailscale Proxy feature
+<b>(click to open)</b></i>
+</summary>
+
+### Home Assistant configuration for Tailscale Proxy feature
```
Comment (frenck) [inline tailscale/DOCS.md:27]: Let's move this section after the installation procedure

File: tailscale/DOCS.md [hunk 9]
```diff
@@ -17,8 +17,55 @@ the following URL:
...
+```
+
+**Note**: _There is no need to adapt anything in these lines since the addon
+runs on your host network._
+
+### Tailscale configuration for Tailscale Proxy feature
+
+[DNS page][tailscale_dns]: Choose a **Tailnet name** and click **Enable HTTPS**
+under HTTPS Certificates (see [Enabling HTTPS][tailscale_info_https] for more
+information)
```
Comment (frenck) [inline tailscale/DOCS.md:66]: This should go after the installation, it is not a pre-requisite.

File: tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run [hunk 10]
```diff
@@ -44,22 +44,70 @@ do
...
-      --advertise-tags="${tags}"
-
-    bashio::exit.ok
-  fi
-  # Well... wait a bit more
   sleep 2
 done
+
+bashio::log.info "Tailscale is running"
+bashio::log.info "Tailscale version: $(/opt/tailscale version --daemon=false | head -n 1)"
```
Comment (frenck) [inline tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:75]: This is already logged

```suggestion
```

File: tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run [hunk 11]
```diff
@@ -44,22 +44,70 @@ do
...
+
+bashio::log.info "Enabling Tailscale Proxy..."
+
+declare domain
```
Comment (frenck) [inline tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:79]: Please declare at the start of the file.

File: tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run
Comment (frenck) [inline tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:1]: Let's extract the proxy stuff out of this file, into its own single shot service

File: tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run [hunk 12]
```diff
@@ -44,22 +44,70 @@ do
...
+
+# Remove previously configured proxy first, because the configuration is persistent between reboots
+/opt/tailscale serve https:443 / "http://localhost:$(bashio::core.port)" off &> /dev/null || true
```
Comment (frenck) [inline tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:82]: IMHO, we should check the current state and make it match and not just turn it off.

File: tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run [hunk 13]
```diff
@@ -44,22 +44,70 @@ do
...
+
+# Check if Tailscale HTTPS is enabled
+if ! /opt/tailscale status --self=true --peers=false --json \
+  | jq -rce ".CertDomains[0]" > /dev/null;
+then
+  bashio::log.info "Can't get your Tailscale domain, is Tailscale's HTTPS support enabled?"
+  bashio::log.info "If you want to use https certificates and https proxy on this device,"
+  bashio::log.info "please change configuration and restart this add-on."
+  bashio::log.info "See the 'Prerequisites' chapter in the add-on's Documentation."
```
Comment (frenck) [inline tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:91]: Instead, we should do nothing.
If it is turned off, we can just log that. All the rest is not needed.

File: tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run [hunk 14]
```diff
@@ -44,22 +44,70 @@ do
...
+  bashio::exit.ok
+fi
+domain=$(/opt/tailscale status --self=true --peers=false --json | jq -rc ".CertDomains[0]")
+
+# Checking if SSL is used
+if bashio::var.true "$(bashio::core.ssl)"; then
+  bashio::log.warning "Tailscale's HTTPS support is enabled, but Home Assistant"
+  bashio::log.warning "is not accessible through plain http connection."
```
Comment (frenck) [inline tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:99]:
```suggestion
  bashio::log.warning "is not accessible through plain HTTP connection."
```

File: tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run [hunk 15]
```diff
@@ -44,22 +44,70 @@ do
...
+  bashio::log.warning "If you wan't to use https certificates and https proxy on this device,"
```
Comment (frenck) [inline tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:100]:
```suggestion
  bashio::log.warning "If you want to use HTTPS certificates and HTTPS proxy on this device,"
```

File: tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run [hunk 16]
```diff
@@ -44,22 +44,70 @@ do
...
+  bashio::log.warning "please change configuration and restart this add-on."
+  bashio::log.warning "See the 'Prerequisites' chapter in the add-on's Documentation."
```
Comment (frenck) [inline tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:102]:
```suggestion
```

**Review by frenck — CHANGES_REQUESTED** [review]

**Comment by lmagyar** [comment]
I've rebased and rewritten as requested.

Also tested manually the proxy startup logic and some corner cases on real devices.

**Comment by MelleD** [comment]
@lmagyar I'm looking forward to the PR 👍

**Comment by frenck** [comment]
@MelleD Not sure, right now, the Home Assistant Core release has my priority this week.

**Comment by lmagyar** [comment]
> Do you think it's possible to do a (beta) release this week?
> I would/could test it over the long Easter weekend :)

@MelleD You can test it with installing my forked repo, more test is better: https://github.com/lmagyar/homeassistant-addon-tailscale

**Comment by MelleD** [comment]
> > Do you think it's possible to do a (beta) release this week?
> > I would/could test it over the long Easter weekend :)
> 
> @MelleD You can test it with installing my forked repo, more test is better: https://github.com/lmagyar/homeassistant-addon-tailscale

Ok great, I'll try that 👍

**Comment by MelleD** [comment]
Short feedback:
Works for me and enabled it. Thanks a bunch

File: tailscale/rootfs/etc/s6-overlay/s6-rc.d/proxy/run [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+  bashio::log.error "Tailscale's HTTPS support is enabled, but Home Assistant is not accessible through plain HTTP connection"
+  bashio::exit.nok
+fi
+
+# Set up proxy
+# Note: Configuration is persistent between reboots, though configuring the same value won't cause duplicate entries,
+#       but will overwite previous entry if core.port has been changed
+# Note: If a tailnet node is renamed, there will be deprecated entries in status json under .Web.<different-domain:443>,
+#       those can't be removed, no command for that situation currently
+if ! /opt/tailscale serve https:443 / "http://127.0.0.1:$(bashio::core.port)"; then
```
Comment (frenck) [inline tailscale/rootfs/etc/s6-overlay/s6-rc.d/proxy/run:39]: So the thing that keeps me from merging this right now, is this line.

It basically sets/dictates it will expose Home Assistant on this port. For most use-cases, this is fine I guess, but I could also think of allowing more customization in the future. For example, use it to expose a completely different service (or additional custom services, for that matter).

I do see that out of the box, this is the wished-for situation though.

🤔

Comment (lmagyar) [reply tailscale/rootfs/etc/s6-overlay/s6-rc.d/proxy/run:39]: Hmmm, something like this? This seems brutal. :) And not supported by the config options schema. :(
```
customize_proxy:
  active: true
  servers:
    - https_port: 443
      funnel: true
      locations:
        - mount_point: /
          local_port: 8123
        - mount_point: /foo/
          local_port: 8124
    - https_port: 8443
      funnel: false
      locations:
        - mount_point: /
          local_port: 8123
        - mount_point: /foo/
          local_port: 8124
```
- If active=false, it serves HA on 443 by default and turns on Funnel.
- Funnel is limited to 443, 8443 and 10000 by Tailscale, we should check it.
- Tailscale serve has memory, so on add-on start it would be better to delete any previous config and recreate it.
- Should I add Funnel functionality to this PR because of the common config?

Comment (frenck) [reply tailscale/rootfs/etc/s6-overlay/s6-rc.d/proxy/run:39]: @Xe I know, I spoke with Shayne already.

Nevertheless, IMHO this is the wrong place to discuss such things and I would appreciate it if you'd not try to sidetrack the subjects of PR reviews for external gains.

Thanks 👍 

../Frenck

**Comment by lmagyar** [comment]
FYI:
- I've squashed and rebased the PR.
- Added a taildrop like optional `bool?` configuration option to disable it in case somebody has any issues with it.

The "how to make it customizable" [review](https://github.com/hassio-addons/addon-tailscale/pull/137#pullrequestreview-1386515841) is still relevant/open.

I also plan to PR the Funnel feature after this PR settled.

File: tailscale/rootfs/etc/s6-overlay/s6-rc.d/proxy/run (hunk 17 above)
Comment (frenck) [reply tailscale/rootfs/etc/s6-overlay/s6-rc.d/proxy/run:39]: @lmagyar Considering the future intentions/changes coming to Tailscale (via their web interface), I think we can just go this route. Once their UI rolls out, we can drop this part.

This keeps us from going into weird YAML configurations as this point.

WDYT?

File: tailscale/translations/en.yaml [hunk 18]
```diff
@@ -29,3 +29,9 @@ configuration:
       This option allows you to enable Taildrop, a file sharing service
       that allows you to share files with other Tailscale nodes.
       When not set, this option is enabled by default.
+  proxy:
+    name: Tailscale Proxy
+    description: >-
+      This option allows you to enable Tailscale's Proxy feature to present your
+      Home Assistant device on your tailnet with a valid certificate.
```
Comment (frenck) [inline tailscale/translations/en.yaml:36]:
```suggestion
      Home Assistant instance on your tailnet with a valid certificate.
```

File: tailscale/DOCS.md [hunk 19]
```diff
@@ -121,6 +122,49 @@ When not set, this option is enabled by default.
 
 Received files are stored in the `/share/taildrop` directory.
 
+### Option: `proxy`
+
+When not set, this option is enabled by default.
+
+Tailscale can provide a TLS certificate for your Home Assistant device within
```
Comment (frenck) [inline tailscale/DOCS.md:129]:
```suggestion
Tailscale can provide a TLS certificate for your Home Assistant instance within
```

File: tailscale/DOCS.md [hunk 20]
```diff
@@ -121,6 +122,49 @@ When not set, this option is enabled by default.
...
+your tailnet domain.
+
+This can prevent browsers to warn that HTTP URLs to your Home Assistant device
+look unencrypted (browsers are not aware of that connections between Tailscale
+nodes are secured with end-to-end encryption).
```
Comment (frenck) [inline tailscale/DOCS.md:134]:
```suggestion
This can prevent browsers from warning that HTTP URLs to your Home Assistant device
look unencrypted (browsers are not aware of the connections between Tailscale
nodes are secured with end-to-end encryption).
```

File: tailscale/DOCS.md [hunk 21]
```diff
@@ -121,6 +122,49 @@ When not set, this option is enabled by default.
...
+
+More information: [Enabling HTTPS][tailscale_info_https]
+
+1. Configure Home Assistant to be accessible through HTTP connection (this is
```
Comment (frenck) [inline tailscale/DOCS.md:138]:
```suggestion
1. Configure Home Assistant to be accessible through an HTTP connection (this is
```

File: tailscale/DOCS.md [hunk 22]
```diff
@@ -121,6 +122,49 @@ When not set, this option is enabled by default.
...
+   the default). See [HTTP integration documentation][http_integration] for more
+   information. If you still want to use another HTTPS connection to access Home
+   Assistant, please use a reverse proxy add-on.
+
+1. Home Assistant, by default, blocks requests from reverse proxies, like the
+   Tailscale Proxy. In order to enable it, add the following lines to your
+   `configuration.yaml`, without changing anything:
```
Comment (frenck) [inline tailscale/DOCS.md:145]:
```suggestion
1. Home Assistant, by default, blocks requests from reverse proxies, like the
   Tailscale Proxy. To enable it, add the following lines to your
   `configuration.yaml`, without changing anything:
```

File: tailscale/DOCS.md [hunk 23]
```diff
@@ -121,6 +122,49 @@ When not set, this option is enabled by default.
...
+
+   - Choose a Tailnet name.
+
+   - Enable MagicDNS if not already enabled.
+
+   - Under HTTPS Certificates section, click Enable HTTPS.
+
+1. Restart the add-on.
+
+**Note:** _You should not use any port number in the url that you used
```
Comment (frenck) [inline tailscale/DOCS.md:164]:
```suggestion
**Note:** _You should not use any port number in the URL that you used
```

File: tailscale/rootfs/etc/s6-overlay/s6-rc.d/proxy/run (hunk 17 above)
Comment (lmagyar) [reply tailscale/rootfs/etc/s6-overlay/s6-rc.d/proxy/run:39]: Yes I agree. I have no better idea even after sleeping a lot on this. :)

**Comment by lmagyar** [comment]
FYI:
- I've accepted everything you suggested
- I've found another "device" -> "instance" replacement in DOCS.md
- Rebased to the latest main branch

**Review by frenck — APPROVED** [review]
Thanks, @lmagyar 👍 

../Frenck
